spring-kafka/src/main/java/org/springframework/kafka/core/KafkaTemplate.java

```diff
@@ -180,22 +179,46 @@ public void setMessageConverter(RecordMessageConverter messageConverter) {
 
 	@Override
 	public List<PartitionInfo> partitionsFor(String topic) {
-		return getTheProducer().partitionsFor(topic);
+		Producer<K, V> producer = getTheProducer();
+		try {
+			return producer.partitionsFor(topic);
+		}
+		finally {
+			producer.close();
+		}
 	}
 
 	@Override
 	public Map<MetricName, ? extends Metric> metrics() {
-		return getTheProducer().metrics();
+		Producer<K, V> producer = getTheProducer();
+		try {
+			return producer.metrics();
+		}
+		finally {
+			producer.close();
+		}
 	}
 
 	@Override
 	public <T> T execute(ProducerCallback<K, V, T> callback) {
-		return callback.doInKafka(getTheProducer());
+		Producer<K, V> producer = getTheProducer();
+		try {
+			return callback.doInKafka(producer);
+		}
+		finally {
+			producer.close();
+		}
 	}
 
 	@Override
 	public void flush() {
```

Comment:
Yeah.... This caused some doubts for me as well.

Yep! Java doc would be great compromise
Thank you!